remotes/docker/pusher.go

```diff
@@ -269,16 +269,15 @@ func (p dockerPusher) push(ctx context.Context, desc ocispec.Descriptor, ref str
 
 	req.body = func() (io.ReadCloser, error) {
 		pr, pw := io.Pipe()
-		pushw.setPipe(pw)
+		pushw.setPipe(pr, pw)
 		return io.NopCloser(pr), nil
 	}
 	req.size = desc.Size
 
 	go func() {
 		resp, err := req.doWithRetries(ctx, nil)
 		if err != nil {
-			pushw.setError(err)
```

Comment:
> Remove the error channel - we don't need this any more, errors are instead clearly tracked through the CloseWithError function.

This could cause a race since the call to `error` could occur when the pipe is still nil. Would keeping the channel but also checking it on Write help the error on Write case?